server.py

```diff
@@ -109,8 +109,33 @@ async def restart_control_server():
         )
     config.register_callback("CONTROL_SERVER_PORT", restart_control_server)
 
-    await instance.listen(("", 8001))
-    await instance.listen(("", 8002), SimpleJsonProtocol)
+    PROTO_CLASSES = {
+        QDataStreamProtocol.__name__: QDataStreamProtocol,
+        SimpleJsonProtocol.__name__: SimpleJsonProtocol
+    }
+    for cfg in config.LISTEN:
+        try:
+            host = cfg.get("address", "")
+            port = cfg.get("port", 0)
+            proto_class_name = cfg.get("protocol", QDataStreamProtocol.__name__)
+            proto_class = PROTO_CLASSES[proto_class_name]
+
+            await instance.listen(
+                address=(host, port),
+                protocol_class=proto_class
+            )
+        except Exception:
+            logger.exception(
+                "Failed to start server instance with config: %s",
+                cfg
+            )
+            done.set_result(-1)
+
+    if not instance.contexts:
+        logger.warning(
+            "The server was not configured to listen on any ports! No one will "
+            "be able to connect."
+        )
```

Comment:
If this is done on startup would it be better to just have it exit with an error so that we know when there are no ports set quickly rather than having to parse the logs

Reply:
Yea I guess maybe it makes more sense to just exit. There is some stuff that could be listening like the control server and the rabbitmq connection, but I'm not sure if there is actually a use case where you'd ever want it to run without allowing clients to connect. Then again, if this does get misconfigured it's not like it's going to go unnoticed since no one will be able to play...

Reply:
Yeah maybe but that requires people complaining ha. And then digging into logs which personally is more of a pain if the service still says it is running fine